wallet/src/Pos/Wallet/Web/Methods/History.hs

```diff
@@ -88,11 +88,12 @@ getHistory mCWalId mAccountId mAddrId = do
         Nothing -> pure accAddrs
         Just addr ->
             if addr `elem` accAddrs then pure [addr] else throwM errorBadAddress
-    first (filter (fits addrs)) <$> getFullWalletHistory cWalId
+    first (filter (fits $ S.fromList addrs)) <$> getFullWalletHistory cWalId
   where
-    fits :: [CId Addr] -> CTx -> Bool
-    fits addrs ctx = any (relatesToAddr ctx) addrs
-    relatesToAddr CTx {..} = (`elem` (map fst $ ctInputs ++ ctOutputs))
+    fits :: S.Set (CId Addr) -> CTx -> Bool
+    fits addrs CTx{..} =
+        let inpsNOuts = map fst (ctInputs ++ ctOutputs)
+        in  any (`S.member` addrs) inpsNOuts
```

Comment:
You don't need to do the set conversion first, just be explicit about the sharing like this:
```
fits :: [CId Addr] -> CTx -> Bool
fits addrs =
    let addrSet = S.fromList addrs -- make the set once, share for many calls with different CTxs
     in \CTx {ctInputs, ctOutputs} ->
          let inoutSet = S.fromList (map fst (ctInputs ++ ctOutputs))
           in intersects addrSet inoutSet
  where
    -- | Do the sets intersect at all?
    intersects :: Set a -> Set a -> Bool
    intersects a b = not (S.null (S.intersection a b))
```